### ISSUE TITLE
Clarification on operator fusion of backpressure operators

### ISSUE THREAD

**Issue by ursusursus** [Kotlin/kotlinx.coroutines#3123]
Hi,
I have library [which uses a Chnnel](https://github.com/cashapp/sqldelight/blob/master/extensions/coroutines-extensions/src/commonMain/kotlin/com/squareup/sqldelight/runtime/coroutines/FlowExtensions.kt#L36)

which uses a Channel(CONFLATED) directly. This means downstream cannot in anyway influence this. Setting makes sense, but I have a use case where I'd like to buffer

If the library got rewritten and used like this
```
callbackFlow
 .conflate() // library because of performance
 .buffer(UNLIMITED) // user space because of certain use case
 ````
 Would this work as expected? I read that `callbackFlow ``uses `BUFFERED channel, and adjacent buffer operators get fused in, so basically downstream controls the original channel behavior, in this case, changes `BUFFERED` to `CONFLATED`

**But how "long" can be the operator fusion?** Can the list be infinite, and the last one wins? Or is there some logic to it?

If its only size of 1, does then `conflate` lose the emits for ever? or would a subsequent `buffer(UNLIMITED)` guarantee to catch them all? or is there some sort of inherent backpressure to `buffer` so conflation could kick in?

**Comment by elizarov** [comment]
The goal of the fusion is to provide a more efficient implementation for a chain of operations that has the same semantics as the original chain. 

One way to understand this `conflate().buffer(UNLIMITED)` combination is that `conflate()` works only for slow consumers. If the consumer is really fast, then `conflate()` has no effect. However, your consumer is `buffer(UNLIMITIED)` and it is a really fast consumer, so a preceding `conflate()` simply has no effect -- it will not have any time to conflate anything, as `buffer(UNLIMITED)` immediately consumes all the emitted items.

Another way to see it is that `conflate()` is equal to `buffer(onBufferOverflow = BufferOverflow.DROP_OLDEST)`. So here we have two buffers, applied one after another. This is inefficient from an implementation standpoint. The fused implementation of two `buffer` applications will optimize it to a single buffer with `buffer(capacity = UNLIMITED, onBufferOverflow = BufferOverflow.DROP_OLDEST)`. However, since the capacity of the buffer is not limited, whatever it does on overflow becomes irrelevant as the buffer will never overflow (until there is any free memory available).

**Comment by ursusursus** [comment]
Okay so if semantics are preserved, this means `conflate().buffer(UNLIMITED)` is not **guaranteed** to receive all emits, it just a best effort, right? i.e. once there is conflation in the chain, then there cannot be a guarantee

Also, could you please comment on this? 

```
val fastFlow = flow {
        repeat(100) {
            emit(it)
        }
    }
    
    
    val withoutBuffer = fastFlow.conflate().onEach { delay(50) }.toList()
    val withBuffer = fastFlow.conflate().buffer(100).onEach { delay(50) }.toList()
    val withMapBuffer = fastFlow.conflate().map { it }.buffer(100).onEach { delay(50) }.toList()
    
    println(withoutBuffer)
    println(withBuffer)
    println(withMapBuffer)
```

https://play.kotlinlang.org/#eyJ2ZXJzaW9uIjoiMS40LjEwIiwiY29kZSI6ImltcG9ydCBrb3RsaW54LmNvcm91dGluZXMuKlxuaW1wb3J0IGtvdGxpbnguY29yb3V0aW5lcy5mbG93LipcblxuZnVuIG1haW4oKSA9IHJ1bkJsb2NraW5nIHtcbiAgICBcbiAgICB2YWwgZmFzdEZsb3cgPSBmbG93IHtcbiAgICAgICAgcmVwZWF0KDEwMCkge1xuICAgICAgICAgICAgZW1pdChpdClcbiAgICAgICAgfVxuICAgIH1cbiAgICBcbiAgICBcbiAgICB2YWwgd2l0aG91dEJ1ZmZlciA9IGZhc3RGbG93LmNvbmZsYXRlKCkub25FYWNoIHsgZGVsYXkoNTApIH0udG9MaXN0KClcbiAgICB2YWwgd2l0aEJ1ZmZlciA9IGZhc3RGbG93LmNvbmZsYXRlKCkuYnVmZmVyKDEwMCkub25FYWNoIHsgZGVsYXkoNTApIH0udG9MaXN0KClcbiAgICB2YWwgd2l0aE1hcEJ1ZmZlciA9IGZhc3RGbG93LmNvbmZsYXRlKCkubWFwIHsgaXQgfS5idWZmZXIoMTAwKS5vbkVhY2ggeyBkZWxheSg1MCkgfS50b0xpc3QoKVxuICAgIFxuICAgIHByaW50bG4od2l0aG91dEJ1ZmZlcilcbiAgICBwcmludGxuKHdpdGhCdWZmZXIpXG4gICAgcHJpbnRsbih3aXRoTWFwQnVmZmVyKVxufSIsInBsYXRmb3JtIjoiamF2YSIsImFyZ3MiOiIifQ==

why does the `withMapBuffer` emits only [0, 99]? is there some inherent overhead to every operator even when it effectively doesnt to anything?

**Comment by elizarov** [comment]
Exactly. Even though `map { it }` between `conflate()` and `buffer()` effectively does not do anything, it still adds some processing. In this case, it receives a value from a previous conflated channel and sends it to the next, buffered channel. Even this small amount of processing is "slow enough" for conflation to start actually conflating.

**Comment by dkhalanskyjb** [comment]
> this means `conflate().buffer(UNLIMITED)` is not guaranteed to receive all emits, it just a best effort, right?

(@elizarov, I think you only answered the second part; please correct me if I'm wrong)
It is guaranteed. `conflate()` drops emissions only if the downstream doesn't process them in time, but `buffer(UNLIMITED)` will process everything immediately, without any queue forming to enter the buffer.

From the implementation standpoint, because we know this, `conflate()` and `buffer(UNLIMITED)` are *fused* together, that is, replaced by a single operation, which doesn't miss any emissions.

**Comment by ursusursus** [comment]
@dkhalanskyjb but I probably should not rely on such impl. detail, or should I? I'm a bit confused, as adding a buffer operator after channelFlow is prescribed in the docs, to tweak the upstream channel setup

**Comment by dkhalanskyjb** [comment]
It's not just an implementation detail, it's documented behavior that we're bound to support. You can always assume, as the documentation says, that `.conflate().buffer(…)` will act as if joined together, without any queue between them, so `.conflate().buffer(UNLIMITED)` will behave the same way as just `.buffer(UNLIMITED)`.

**Comment by dkhalanskyjb** [comment]
Well, this has nothing to do with us, right? It's up to each library to decide on their own stability guarantees. There's no real guarantee that the library we're discussing (for the record, I don't know which one it is, so these are not some personal gripes) will be providing the function that produces that flow at all.

If you ask me, my personal rule of thumb when using libraries is to rely on the behavior that is documented. If `.conflate()` being the last operator is not documented, probably don't rely on that, or ask the authors for clarifications.

For example, when writing `ObservableValue<T>.asFlow()`, which is implemented with `.conflate()`, here's the description that we came up with: https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-javafx/kotlinx.coroutines.javafx/as-flow.html So, in this case, you have a guarantee.

**Comment by joffrey-bion** [comment]
@dkhalanskyjb Sorry if my previous comment came out wrong. I completely agree this has nothing to do with kotlinx.coroutines, I was just trying to help @ursusursus for his actual use case. I linked the initial Slack thread above for reference.

> If you ask me, my personal rule of thumb when using libraries is to rely on the behavior that is documented.

I totally agree here too. This is the definition of "breaking change" that I like to use: the contract is the doc.
